### PR TITLE
plabuilder: use OR for not in comparisons

### LINKED ISSUE #14605: Bug Report: Incorrect Results Returned by Cross-Shard Subquery in NOT IN Clause
### Overview of the Issue

There is an SQL query: ``select id from user where id not in (select col from user)``. This SQL statement returns incorrect results.

### Reproduction Steps

1.Deploy the following vschema:
```
{
    "sharded": true,
    "vindexes": {
        "hash": {
            "type": "hash"
        },
        "tables": {
            "user": {
                "column_vindexes": [
                    {
                        "column": "Id",
                        "name": "user_index"
                    },
                    {
                        "column": "Name",
                        "name": "name_user_map"
                    },
                    {
                        "column": "Costly",
                        "name": "costly_map"
                    }
                ],
                "auto_increment": {
                    "column": "id",
                    "sequence": "seq"
                },
                "columns": [
                    {
                        "name": "col",
                        "type": "INT16"
                    },
                    {
                        "name": "predef1"
                    },
                    {
                        "name": "predef2"
                    },
                    {
                        "name": "textcol1",
                        "type": "VARCHAR"
                    },
                    {
                        "name": "intcol",
                        "type": "INT16"
                    },
                    {
                        "name": "textcol2",
                        "type": "VARCHAR"
                    }
                ]
            }
        }
    }
}
```

2.View error
In the [filter_cases.json](https://github.com/vitessio/vitess/blob/main/go/vt/vtgate/planbuilder/testdata/filter_cases.json#L1935) file, it will be rewritten into two SQL queries: 

```sql
select col from user 
select id from user where not :__sq_has_values and id not in ::__sq1. 
```

When the select col from user query returns one result, the value of `__sq_has_values` becomes 1. Consequently, the where condition in the second SQL query becomes not 1, resulting in:

 ```sql
select id from user where not 1 and id not in ::__sq1
```

However, a condition like where not 1 will return an Empty set.

### Binary Version

```sh
vtgate version Version: 18.0.1 (Git revision aa72dc8d9189843c0a49f32b54f6fc978f6dbcde branch 'HEAD') built on Mon Nov 27 15:09:22 CST 2023 by wangwei@ZBMac-xxxxx using go1.21.4 darwin/arm64
```

### Operating System and Environment details

```sh
Darwin 21.1.0
arm64
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by systay**
## Description
The plans produced for `NOT IN` queries that have to be scattered did not handle the missing values case, which leads to wrong results for these situations.

## Related Issue(s)
Fixes #14605

## Checklist0
-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — APPROVED** [review]

**Review by arthurschreiber — APPROVED** [review]
😢

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vtgate/planbuilder/testdata/filter_cases.json [hunk 1]
```diff
@@ -1966,7 +1966,7 @@
               "Sharded": true
             },
             "FieldQuery": "select id from `user` where 1 != 1",
-            "Query": "select id from `user` where not :__sq_has_values and id not in ::__sq1",
+            "Query": "select id from `user` where not :__sq_has_values or id not in ::__sq1",
```
Comment (wangweicugw) [inline go/vt/vtgate/planbuilder/testdata/filter_cases.json:1969]: I have a question. It seems that in this scenario, we can remove the "not :__sq_has_values" condition. Why do we need to convert it to "or"?

Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/filter_cases.json:1969]: Excellent question!

We have already built the queries by this time, so we need to be able to send down something that is valid SQL. 
`NOT IN ()` does not even parse - we need to have at least one value there. 
We can't stick `NULL` there, `NOT IN (NULL)` evaluates to `NULL`, not `true`.
We just put in the fake value 0. Just to make sure that we end up with `true` for the empty results, even if the column we are comparing against does contain a `0`, we add this `not :__sq_has_values` that will return true no matter what.

Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/filter_cases.json:1969]: If it is important, we would need to come up with a value that we put into the `::__sq1` tuple that guarantees that `id not in ::__sq1` returns true for all values of `id`. We can make it very unlikely, but not guarantee that we will no clash, right?

**Pull request merged**